bn_mp_kronecker.c

```diff
@@ -0,0 +1,141 @@
+#include "tommath_private.h"
+#ifdef BN_MP_KRONECKER_C
+
+/* LibTomMath, multiple-precision integer library -- Tom St Denis
+ *
+ * LibTomMath is a library that provides multiple-precision
+ * integer arithmetic as well as number theoretic functionality.
+ *
+ * The library was designed directly after the MPI library by
+ * Michael Fromberger but has been written from scratch with
+ * additional optimizations in place.
+ *
+ * The library is free for all purposes without any express
+ * guarantee it works.
+ */
+
+/*
+   Kronecker symbol (a|p)
+   Straightforward implementation of algorithm 1.4.10 in
+   Henri Cohen: "A Course in Computational Algebraic Number Theory"
+
+   @book{cohen2013course,
+     title={A course in computational algebraic number theory},
+     author={Cohen, Henri},
+     volume={138},
+     year={2013},
+     publisher={Springer Science \& Business Media}
+    }
+ */
+int mp_kronecker(const mp_int *a, const mp_int *p, int *c)
+{
+   mp_int a1, p1, r;
+
+   int e = MP_OKAY;
+   int v, k;
+
+   const int table[8] = {0, 1, 0, -1, 0, -1, 0, 1};
+
+   if (mp_iszero(p)) {
+      if (a->used == 1 && a->dp[0] == 1) {
```

Comment:
Re: helper-pl results: I repaired the white-space etc. but what about the static functions? Not more than one function per file or no `static` functions at all? Both can be replaced with a rather ugly and barely readable macro if necessary, but is it?

The free'ing in `mp_kronecker` was a mess, indeed, thank you for pointing that out!

Testing for failures (although mainly allocation but others are possible, too) in libtommath is quite tedious. It would be really nice to have something to automate it. But that is just a "nice-to-have" and goes to the wishlist for 2020 ;-)

Re: using a macro for the `issmalldigit` kind of test&hellip;I do not think it would be a good idea because the digit `d` needs to be smaller than or equal to `MP_MASK`. That needs to be checked and I would advise to use `mp_cmp_d()` in such cases. On the other hand: a macro like
```c
#define  mp_isunity(a) (((a)->used == 1) && ((a)->dp[0] == 1))
```
might be useful but I already hear the complaints that it returns true for `-1`, too \*sigh\*.

My use here is correct because `kronecker(1,0) = kronecker(-1,0) = 1` and `kronecker(x,0) = 0` for all other `x` and a function call can be expensive. It would be no problem to use `mp_cmp_d()` instead if you so wish.

It is all locally ready for a `push` to this fork with the exception of the `static` functions.

Reply:
>  but what about the static functions?

ah, we decided at one point that static functions should be prefixed with `_`

> wishlist for 2020

I appreciate your over-optimism ;-)

> My use here is correct because...

fine by me then as it is!

Reply:
> ah, we decided at one point that static functions should be prefixed with `_`

I really should have looked into the source of `helper.pl`, yes ;-)
But no problem, done.

No I get `first_comment` and `last_comment`
```
[first_comment] tommath_class.h
[last_comment] tommath_class.h
[last_comment] bn_mp_get_bit.c
[last_comment] bn_mp_kronecker.c
[last_comment] bn_mp_prime_frobenius_underwood.c
[last_comment] bn_mp_prime_strong_lucas_selfridge.c
check-comments:  FAIL 6
```
I added them manually to ` tommath_class.h` for now, should be put in the script in the near future. The rest passes now.
Will `push` now, let's see what comes up this time ;-)
(Oh, and please check the free'ings  in `kronecker.c`, thanks)